internal/benchrunner/runners/rally/report.go

```diff
@@ -0,0 +1,239 @@
+
+package rally
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/jedib0t/go-pretty/table"
+	"github.com/jedib0t/go-pretty/text"
+
+	"github.com/elastic/elastic-package/internal/benchrunner/reporters"
+	"github.com/elastic/elastic-package/internal/elasticsearch/ingest"
+)
+
+type report struct {
+	Info struct {
+		Benchmark            string
+		Description          string
+		RunID                string
+		Package              string
+		StartTs              int64
+		EndTs                int64
+		Duration             time.Duration
+		GeneratedCorporaFile string
+	}
+	Parameters struct {
+		PackageVersion   string
+		DataStream       dataStream
+		Input            string
+		Vars             map[string]interface{}
+		WarmupTimePeriod time.Duration
+		Corpora          corpora
+	}
+	ClusterName         string
+	Nodes               int
+	DataStreamStats     *ingest.DataStreamStats
+	IngestPipelineStats map[string]ingest.PipelineStatsMap
+	DiskUsage           map[string]ingest.DiskUsage
+	TotalHits           int
+	RallyStats          []rallyStat
+}
+
+func createReport(benchName, corporaFile string, s *scenario, sum *metricsSummary, stats []rallyStat) (reporters.Reportable, error) {
+	r := newReport(benchName, corporaFile, s, sum, stats)
+	human := reporters.NewReport(s.Package, reportHumanFormat(r))
+
+	jsonBytes, err := reportJSONFormat(r)
+	if err != nil {
+		return nil, fmt.Errorf("rendering JSON report: %w", err)
+	}
+
+	jsonFile := reporters.NewFileReport(s.Package, fmt.Sprintf("system/%s/report.json", sum.RunID), jsonBytes)
+
+	mr := reporters.NewMultiReport(s.Package, []reporters.Reportable{human, jsonFile})
+
+	return mr, nil
+}
+
+func newReport(benchName, corporaFile string, s *scenario, sum *metricsSummary, stats []rallyStat) *report {
+	var report report
+	report.Info.Benchmark = benchName
+	report.Info.Description = s.Description
+	report.Info.RunID = sum.RunID
+	report.Info.Package = s.Package
+	report.Info.StartTs = sum.CollectionStartTs
+	report.Info.EndTs = sum.CollectionEndTs
+	report.Info.Duration = time.Duration(sum.CollectionEndTs-sum.CollectionStartTs) * time.Second
+	report.Info.GeneratedCorporaFile = corporaFile
+	report.Parameters.PackageVersion = s.Version
+	report.Parameters.Input = s.Input
+	report.Parameters.Vars = s.Vars
+	report.Parameters.DataStream = s.DataStream
+	report.Parameters.WarmupTimePeriod = s.WarmupTimePeriod
+	report.Parameters.Corpora = s.Corpora
+	report.ClusterName = sum.ClusterName
+	report.Nodes = sum.Nodes
+	report.DataStreamStats = sum.DataStreamStats
+	report.IngestPipelineStats = sum.IngestPipelineStats
+	report.DiskUsage = sum.DiskUsage
+	report.TotalHits = sum.TotalHits
+	report.RallyStats = stats
+	return &report
+}
+
+func reportJSONFormat(r *report) ([]byte, error) {
+	b, err := json.MarshalIndent(r, "", "\t")
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func reportHumanFormat(r *report) []byte {
```

Comment:
Trying to get my head around how this report exactly works. What happens if rally adds new fields to the report, does it mean we have to update our code? Does rally itself already have some tooling / code in place to output it into a human readable format?

Reply:
this is basically a copy of the report that @marc-gr created for system benchmark.
I just added the stats from rally

rally outputs to stdout in human readable format by default, but the output is mixed with the rest, so I save a report from rally in csv format, and dump here with our pretty table :)

since the source is a csv no matter if rally adds new fields, we'll get them with no change required, as long as they are saved in the rally csv repo